src/frontend/xgboost/xgboost_json.h

```diff
@@ -0,0 +1,376 @@
+
+  bool StartObject();
+
+  bool String(const char *str, std::size_t length, bool copy);
+};
+
+/*! \brief handler for LearnerParam objects from XGBoost schema*/
+class LearnerParamHandler : public OutputHandler<treelite::Model> {
+ public:
+  using OutputHandler<treelite::Model>::OutputHandler;
+  bool String(const char *str, std::size_t length, bool copy);
+};
+
+/*! \brief handler for Learner objects from XGBoost schema*/
+class LearnerHandler : public OutputHandler<treelite::Model> {
+ public:
+  using OutputHandler<treelite::Model>::OutputHandler;
+  bool StartArray();
+  bool StartObject();
+  bool EndObject(std::size_t memberCount);
+
+ private:
+  std::string objective;
+};
+
+/*! \brief handler for XGBoostModel objects from XGBoost schema*/
+class XGBoostModelHandler : public OutputHandler<treelite::Model> {
+ public:
+  using OutputHandler<treelite::Model>::OutputHandler;
+  bool StartArray();
+  bool StartObject();
+  bool EndObject(std::size_t memberCount);
+
+ private:
+  std::vector<unsigned> version;
+};
+
+/*! \brief handler for root object of XGBoost schema*/
+class RootHandler : public OutputHandler<treelite::Model> {
+ public:
+  using OutputHandler<treelite::Model>::OutputHandler;
+  bool StartObject();
+};
+
+/*! \brief handler which delegates JSON parsing to stack of delegates*/
+class DelegatedHandler
+    : public rapidjson::BaseReaderHandler<rapidjson::UTF8<>, DelegatedHandler>,
+      public Delegator {
+
+ public:
+  /*! \brief create DelegatedHandler with initial RootHandler on stack */
+  static std::shared_ptr<DelegatedHandler> create() {
+    struct make_shared_enabler : public DelegatedHandler {};
+
+    std::shared_ptr<DelegatedHandler> new_handler = \
+      std::make_shared<make_shared_enabler>();
+    new_handler->push_delegate(std::make_shared<RootHandler>(
+      new_handler,
+      new_handler->result));
+    return new_handler;
+  }
+
+  /*! \brief push new handler onto stack, delegating ongoing parsing to it
+   *  \param new_delegate the delegate to push onto stack
+   */
+  void push_delegate(
+      std::shared_ptr<BaseHandler> new_delegate) override {
+    delegates.push(new_delegate);
+  }
+  /*! \brief pop handler off of stack, returning parsing responsibility to
+   *         previous handler on stack
+   */
+  void pop_delegate() override {
+    delegates.pop();
+  }
+  treelite::Model get_result();
+  bool Null();
+  bool Bool(bool b);
+  bool Int(int i);
+  bool Uint(unsigned u);
+  bool Int64(int64_t i);
+  bool Uint64(uint64_t u);
+  bool Double(double d);
+  bool String(const char *str, std::size_t length, bool copy);
+  bool StartObject();
+  bool Key(const char *str, std::size_t length, bool copy);
+  bool EndObject(std::size_t memberCount);
+  bool StartArray();
+  bool EndArray(std::size_t elementCount);
+
+ private:
+  DelegatedHandler() : delegates{}, result{} {};
+
+  std::stack<std::shared_ptr<BaseHandler>> delegates;
+  treelite::Model result;
```

Comment:
```suggestion
  std::stack<std::shared_ptr<BaseHandler>> m_delegates;
  treelite::Model m_result;
```
to be consistent with `m_output`